### PR TITLE
Initial implementation of a polling event loop for Node.js

### PR DESCRIPTION
**Pull request by tronical**
Todos:

 - [x] Use `pump_events` with winit instead of `quit_event_loop()` from `invoke_from_event_loop()` to distinguish between quitting due to last window closed or just finishing the pumping.
 - ~~[ ] Add support for `process_events` to linuxkms~~ -> separate PR.
 - [x] Add support for `process_events` to qt
 - ~~[ ] Add support for poll-free event loop on !windows using libuv's backend fd.~~ -> separate PR.
 - [x] Improve docs on nodejs side for event loop
 - [x] Improve docs in Rust platform API for process events

### CODE REVIEW

File: internal/backends/winit/lib.rs [hunk 1]
```diff
@@ -233,7 +234,10 @@ impl i_slint_core::platform::Platform for Backend {
     }
 
     fn run_event_loop(&self) -> Result<(), PlatformError> {
-        crate::event_loop::run()
+        let loop_state = self.event_loop_state.borrow_mut().take().unwrap_or_default();
+        let (new_state, res) = loop_state.run();
```
Comment (ogoffart) [inline internal/backends/winit/lib.rs:238]: should run perhaps take a &mut self?
So you can do
```rust
self.event_loop_state.borrow_mut().get_or_insert_with(Default::default).run()
```

File: api/napi/src/interpreter/component_instance.rs [hunk 2]
```diff
@@ -36,8 +36,19 @@ impl JsComponentInstance {
     }
 
     #[napi]
-    pub fn run(&self) {
-        self.inner.run().unwrap()
+    pub fn show(&self, env: Env) -> Result<JsUndefined> {
```
Comment (ogoffart) [inline api/napi/src/interpreter/component_instance.rs:39]: Why is this needed?  this doesn't seem to be called as the TS does `this.instance.window().show();`  (note the `window()` )

Comment (tronical) [reply api/napi/src/interpreter/component_instance.rs]: Well spotted - thanks!

File: internal/backends/winit/lib.rs (hunk 1 above)
Comment (tronical) [reply internal/backends/winit/lib.rs:238]: I had the same thought, but it doesn't compile with the wasm build (that uses run() instead of run_return()):

```
error[E0521]: borrowed data escapes outside of method
   --> internal/backends/winit/event_loop.rs:701:13
    |
644 |       pub fn run(&mut self) -> Result<(), corelib::platform::PlatformError> {
    |                  ---------
    |                  |
    |                  `self` is a reference that is only valid in the method body
    |                  let's call the lifetime of this reference `'1`
...
701 | /             winit_loop.run(
702 | |                 move |event: Event<SlintUserEvent>,
703 | |                       event_loop_target: &EventLoopWindowTarget<SlintUserEvent>,
704 | |                       control_flow: &mut ControlFlow| {
...   |
712 | |                 },
713 | |             )
    | |             ^
    | |             |
    | |_____________`self` escapes the method body here
    |               argument requires that `'1` must outlive `'static`
    |
    = note: requirement occurs because of the type `RefCell<(Box<dyn ClipboardProvider>, Box<dyn ClipboardProvider>)>`, which makes the generic argument `(Box<dyn ClipboardProvider>, Box<dyn ClipboardProvider>)` invariant
    = note: the struct `RefCell<T>` is invariant over the parameter `T`
    = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
```

**Comment by tronical** [comment]
Changing this back to be a draft. This needs a bit more work, apart from rebasing. I'll push some work in progress for CI testing, but I'll mark it as read for review when it's ready.

**Comment by tronical** [comment]
I'm taking this out of draft mode as I'd like to get feedback on this. If this seems like a reasonable approach, then I can do the other missing TODOs listed in the PR description as follow-up changes.

**Comment by tronical** [comment]
For review I recommend going through the individual commits, as otherwise the entire diff may be harder to read - at least the winit bits :)

File: api/node/Cargo.toml [hunk 3]
```diff
@@ -23,10 +23,13 @@ napi = { version = "2.12.0", default-features = false, features = ["napi8"] }
 napi-derive = "2.12.2"
 i-slint-compiler = { workspace = true, features = ["default"] }
 i-slint-core = { workspace = true, features = ["default"] }
+i-slint-backend-selector = { workspace = true }
 slint-interpreter = { workspace = true, features = ["default", "display-diagnostics", "internal"] }
+slint = { workspace = true, features = ["default"] }
```
Comment (ogoffart) [inline api/node/Cargo.toml:28]: Why is that needed?   it seems the item you try to access are in i-slint-core or i-slint-backend-selector

But if you really need that, i don't think you should enable the default features here.

File: api/node/__test__/eventloop.spec.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+test.serial('merged event loops with timer', async (t) => {
+
+    let invoked = false;
+
+    await run_event_loop(() => {
+        invoked = true;
+
+        setTimeout(() => {
+            quit_event_loop();
```
Comment (ogoffart) [inline api/node/__test__/eventloop.spec.ts:21]: Could you set invoked here instead?

File: api/node/__test__/eventloop.spec.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        }, 2);
+    });
+    t.true(invoked)
+})
+
+
+test.serial('merged event loops with networking', async (t) => {
+    const port = 8000;
```
Comment (ogoffart) [inline api/node/__test__/eventloop.spec.ts:29]: It would be better not to hardcode the port here.
You can call [`server.listen`](https://nodejs.org/api/net.html#serverlistenport-host-backlog-callback) without a port:

> If port is omitted or is 0, the operating system will assign an arbitrary unused port, which can be retrieved by using server.address().port after the ['listening'](https://nodejs.org/api/net.html#event-listening) event has been emitted.

So the test works if there are other test running in parallel that would like to use that port number. (hopefully there isn't because no tests should hardcode ports :-))

File: api/node/index.ts [hunk 6]
```diff
@@ -438,12 +440,68 @@ export function loadFile(filePath: string, options?: LoadFileOptions): Object {
...
+
+        let id = setInterval(() => {
+            if (napi.processEvents() == napi.ProcessEventsResult.Exited || this.#quit_loop) {
+                clearInterval(id);
+                this.#terminate_resolve_fn!(undefined);
+                this.#terminate_resolve_fn = null;
+                this.#termination_promise = null;
+                return;
+            }
+        }, 16);
```
Comment (ogoffart) [inline api/node/index.ts:475]: > 16

magic number?
Yes, i know that it is 1000/60
maybe deserve a comment.

**Comment by ogoffart** [comment]
In the commit title:

> [winit: Clean up event loop state handling part 1](https://github.com/slint-ui/slint/pull/3718/commits/282e2a72e16126c4f62ccfe369b23b335728008d)
> [winit: Clean up event loop state handling part 1](https://github.com/slint-ui/slint/pull/3718/commits/c0b6fe09e1eaa82187a18f3be40b75bcfa757382)

One of them ought to be part 2 :-)

File: api/node/Cargo.toml (hunk 3 above)
Comment (tronical) [reply api/node/Cargo.toml:28]: Good point, I'll try to eliminate the dependency.

File: api/node/__test__/eventloop.spec.ts (hunk 4 above)
Comment (tronical) [reply api/node/__test__/eventloop.spec.ts:21]: Good idea!

File: api/node/__test__/eventloop.spec.ts (hunk 5 above)
Comment (tronical) [reply api/node/__test__/eventloop.spec.ts:29]: Good point. Done. There's one caveat: There might be prompts resulting from this. When running this on macOS I get one. But as long as it works in the CI that's fine.

File: api/node/__test__/eventloop.spec.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        result.writeHead(200);
+        result.end("Hello World");
+    };
+
+    let received_response = "";
+
+    await run_event_loop(() => {
+
+        const server = http.createServer(listener);
+        server.listen(port, host, async () => {
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/node/__test__/eventloop.spec.ts]: Good point, I'll add a test for that.

[the remark this answers is not in the document]
Comment (tronical) [reply api/node/__test__/eventloop.spec.ts]: On second thought, this is what's happening right now:

1. We schedule the callback provided to `run_event_loop()` to be invoked by calling `invoke_from_event_loop()`.
2. The Slint event loop is started.
3. The callback here is invoked, `server.listen()` is called with a callback. That callback is not invoked immediately, it's invoked in the next nodejs loop iteration.
3. After the call to `listen()` we therefore continue with the Slint event loop, and then we return to the nodejs loop.
4. Latest by now nodejs called the real listen() system call, and then invokes the callback here (line 42).
5. This starts a http fetch and chains promises, then the timeout handler from `index.ts` gets called, we re-enter Slint's event loop. And then back to nodejs.

How would you re-order this to improve coverage?

File: api/node/index.ts [hunk 8]
```diff
@@ -438,12 +440,68 @@ export function loadFile(filePath: string, options?: LoadFileOptions): Object {
...
-    export function singleShot(duration: number, handler: () => void) {
-        napi.singleshotTimer(duration, handler)
+class EventLoop {
+    #quit_loop: boolean = false;
+    #termination_promise: Promise<unknown> | null = null;
+    #terminate_resolve_fn: ((_value: unknown) => void) | null;
+    constructor() {
+    }
+
+    start(running_callback?: Function) {
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/node/index.ts:450]: The callbacks return value is ignored. Should we do something with it?

File: api/node/index.ts (hunk 6 above)
Comment (tronical) [reply api/node/index.ts:475]: You're right.

File: api/node/index.ts [hunk 9]
```diff
@@ -438,12 +440,68 @@ export function loadFile(filePath: string, options?: LoadFileOptions): Object {
...
+        if (this.#termination_promise != null) {
+            return this.#termination_promise;
+        }
+
+        this.#termination_promise = new Promise((resolve) => {
+            this.#terminate_resolve_fn = resolve;
+        });
+        this.#quit_loop = false;
+
+        if (running_callback! != undefined) {
```
[the remark this answers is not in the document]
Comment (tronical) [reply api/node/index.ts:460]: Well spotted, the `!` is only for the type checker, but it doesn't make sense here :)

File: internal/backends/qt/qt_widgets.rs [hunk 10]
```diff
@@ -154,9 +154,12 @@ cpp! {{
     #include <QtCore/QMimeData>
     #include <QtCore/QDebug>
     #include <QtCore/QScopeGuard>
+    #include <iostream>
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/backends/qt/qt_widgets.rs:157]: Yes, leftover. Thanks :)

File: internal/backends/qt/qt_widgets.rs [hunk 11]
```diff
@@ -230,6 +233,20 @@ cpp! {{
         ptr->animation_update_property_ptr = animation_update_property_ptr;
         return ptr;
     }
+
+    struct EventSpy : public QObject {
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/backends/qt/qt_widgets.rs]: Ooops, also a leftover.

File: internal/core/platform.rs [hunk 12]
```diff
@@ -99,6 +107,15 @@ pub trait Platform {
     }
 }
 
+/// This enum describes the two possible outcomes of calling [`Platform::process_events`].
+pub enum ProcessEventsResult {
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/core/platform.rs]: `ControlFlow` is an excellent idea. Less is more. Thanks :)

File: api/node/index.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (ogoffart) [reply api/node/index.ts:450]: I thought it'd return a promise of some sort since there is a bunch of return

[the remark this answers is not in the document]
Comment (tronical) [reply api/node/index.ts:450]: We could require the user-provided running callback to return a promise, but what would the meaning be? Should we wait until that promise is resolved, before continuing with the event loop? Not sure what the use-case would be.

As to the return value of `run_event_loop()`/`run()`: That's indeed a promise. We could make it resolve to whatever this callback returns, but I'm also not sure what the use-case would be. This callback is invoked at the start, so why should the event loop at a much later return what this startup "notifier" returned.

Either way, I'll work on the docs of the user-facing API :)

File: internal/core/platform.rs [hunk 13]
```diff
@@ -36,6 +36,14 @@ pub trait Platform {
         Err(PlatformError::NoEventLoopProvider)
     }
 
+    /// Spins an event loop for a specified period of time.
```
[the remark this answers is not in the document]
Comment (tronical) [reply internal/core/platform.rs:39]: It's true that at least in winit many implementations of `run_on_demand()` (our `run_event_loop()`) basically call `pump_events()` (this `process_events()`). But it's not 100% the same, for example on x11 there's what looks like an extra `XSync()` call, and on macOS there's some different about stopping the NSApp after receiving `draw_rect()` (redraw request).

At this point I'd like to play it safe and not merge the two, to reduce the risk of regressions. Instead, I'll mark it as `#[doc(hidden)]` so that we have options to change it in the future. The same applies to the return value, maybe we'll find out that we need something in addition.

**Comment by tronical** [comment]
I'll rebase this now to resolve conflicts.

**Review by ogoffart — APPROVED** [review]
